### ISSUE TITLE
Successor #824/#3458: test rapid Cactus Needle 3 skill adaptation from Astra demonstrations

### ISSUE THREAD

**Issue by Unjuno** [Unjuno/agent-interface#4205]
## Lineage and distinction

Parents / relevant predecessors:

- #9 — broad ephemeral learned local control from Astra demonstrations/corrections.
- #824 — completed sparse-label task-local policy compilation using KNN/linear/tree on synthetic structured decisions.
- #3458 — intent-preserving local distillation research.
- #3441 — adaptive parameter updates and per-skill routing evidence.
- #4158 — current provisional delegation guidance.

This Issue concerns **Cactus Compute Needle 3** as a concrete small structured-action model. It is not a rerun of #824 and does not replace #3458.

The new discriminator is:

> Can a pre-trained 29–121M-class structured-action model be adapted from a small Astra demonstration/correction set quickly enough to become a session/skill-local action delegate, while preserving a versioned immutable active snapshot and explicit YIELD?

Current-main intake: `5542d80583616b204531e1eb590125a41704929c`.

## H — hypothesis

Starting from one pinned Cactus Needle 3 checkpoint, a small task-local fine-tune/adapter built from a frozen small set of Astra-style demonstrations can improve held-out structured-action/effect fidelity over the base checkpoint within a practically useful update time, without silently overwriting the currently active skill snapshot.

The useful outcome is not "training succeeded." It requires measurable held-out gain, preserved scope/YIELD behavior, versioned switch/rollback integrity, and acceptable update cost.

## T — smallest meaningful adaptation experiment

Authority-neutral / owned replay fixture first. No user desktop action.

Prerequisite: a frozen structured-action task from the Cactus Needle 3 action-delegate Issue or an equivalent already retained task contract.

Freeze:

- exact base checkpoint/depth/runtime;
- tool schema;
- intent/state representation;
- one fixed held-out evaluation set;
- demonstration/support pool;
- adaptation recipe/hyperparameters;
- model/adapter snapshot format;
- activation/switch protocol;
- decision gates before any formal fitting.

Compare support budgets without tuning after the first formal outcome. Candidate budgets:

```text
0 demonstrations  = frozen base
16 demonstrations
32 demonstrations
64 demonstrations
```

If four budgets make one formal allocation too broad, begin with exactly `0 vs 32` and leave the others to successors.

Examples must contain only observable Astra-style supervision:

```text
current state
intent / constraints
Astra-approved structured action
next observation / verified effect where available
correction or YIELD label
```

No hidden chain-of-thought.

## Snapshot discipline

The currently active snapshot remains immutable during candidate training:

```text
ACTIVE_A
   |
   +-- demonstrations/corrections --> CANDIDATE_B
                                      |
                                      +-- shadow held-out validation
                                             |
                                   PASS -----+----- FAIL
                                    |                 |
                               safe switch        discard
                                    |
                               ACTIVE_B
```

Old-version proposals arriving after switch must be rejected. Rollback restores model state only; it does not undo already produced application effects.

## Measurements

- adaptation wall time;
- peak memory / accelerator requirement;
- adapter/package bytes;
- held-out semantic tool/action accuracy;
- independently acceptable effect rate;
- forbidden-effect proposal rate;
- required-YIELD recall;
- same-state/different-intent response where applicable;
- base -> tuned gain;
- base/general-skill regression on a frozen control set;
- warm inference p50/p95 before vs after tuning;
- snapshot round-trip;
- version switch and stale-proposal rejection;
- rollback exactness.

## D — decisions

- `PASS_CACTUS_NEEDLE3_RAPID_SKILL_ADAPTATION_SCOPED` — candidate gains the preregistered held-out amount over base, meets all effect/YIELD/scope gates, and stays inside the frozen update-time/resource budget.
- `FAIL_NO_USEFUL_ADAPTATION_GAIN` — update completes but held-out gain is insufficient.
- `FAIL_SKILL_INTERFERENCE` — adaptation materially damages the frozen control skill/general envelope.
- `FAIL_YIELD_OR_SCOPE_REGRESSION` — tuned candidate becomes more willing to act outside the declared envelope.
- `FAIL_SNAPSHOT_SWITCH_INTEGRITY` — stale/version/switch/rollback semantics fail.
- `HOLD_UPDATE_TOO_SLOW_FOR_SESSION_LOCAL_USE` — quality passes but update time is outside the declared use case.
- `STOP_MODEL_OR_TRAINING_PROVENANCE` — exact artifacts cannot be frozen.

Do not reclassify an update-quality FAIL as success because training is fast.

## C — competing explanations

- A small deterministic/tree policy may learn the same skill faster from the same demonstrations.
- Improvement may be memorization rather than state/intent generalization.
- Pretraining may not help enough to justify the heavier runtime.
- Fine-tuning may shift YIELD/abstention behavior more than task competence.
- 32/64 demonstrations may be unrealistically expensive in Astra teacher time.

## U — uncertainty / limits

- First rung is one skill family only.
- No claim of continuous per-frame online learning.
- "Real-time learning" here means **asynchronous/session-local candidate construction**, not weight updates inside every 20–60 ms control cycle.
- No live task authority follows from shadow adaptation.
- If adaptation requires minutes or large accelerators, Needle may still be a useful static skill model but not the intended rapid delegate.

## Successor

Only after a scoped adaptation PASS should a new Issue test live closed-loop DAgger-style correction collection or depth/speculative acceleration. Preserve #824/#3441/#3458 negative evidence unchanged.
